### PR TITLE
fixed issue regarding concurrent requests to dns

### PR DESCRIPTION
**Pull request by esnunes**

### CODE REVIEW

**Comment by davglass** [comment]
Nice catch!

**Comment by davglass** [comment]
@drewfish @sylvio

**Comment by yahoocla** [comment]
CLA is valid!

File: lib/index.js [hunk 1]
```diff
@@ -94,23 +100,29 @@ var EnhanceDns = function (conf) {
...
+                list.push(callback);
+
+                if (list.length > 1) { return; }
+
+                backup_object.lookup(domain, family, function (err, address, family_r) {
+                    if (err) {
+                        list.forEach(function (cb) { cb(err); });
+                        delete callbackList[key];
+                    } else {
+                        cache.set('lookup_' + domain + '_' + family, {
```
Comment (drewfish) [inline lib/index.js]: Could you use the `key` variable here too? If we change the way the key is constructed it'd be nice to only do it in one place.

File: lib/index.js [hunk 2]
```diff
@@ -126,172 +138,226 @@ var EnhanceDns = function (conf) {
...
+                list.push(callback_new);
+
+                if (list.length > 1) { return; }
+
+                backup_object.resolve(domain, type_new, function (err, addresses) {
+                    if (err) {
+                        list.forEach(function (cb) { cb(err); });
+                        delete callbackList[key];
+                    } else {
+                        cache.set('resolve_' + domain + '_' + type_new, addresses, function () {
```
Comment (drewfish) [inline lib/index.js]: Could you use the `key` variable here too? If we change the way the key is constructed it'd be nice to only do it in one place.

File: lib/index.js [hunk 3]
```diff
@@ -126,172 +138,226 @@ var EnhanceDns = function (conf) {
...
+                list.push(callback);
+
+                if (list.length > 1) { return; }
+
+                backup_object.resolve4(domain, function (err, addresses) {
+                    if (err) {
+                        list.forEach(function (cb) { cb(err); });
+                        delete callbackList[key];
+                    } else {
+                        cache.set('resolve4_' + domain, addresses, function () {
```
Comment (drewfish) [inline lib/index.js]: Could you use the `key` variable here too? If we change the way the key is constructed it'd be nice to only do it in one place.

File: lib/index.js [hunk 4]
```diff
@@ -126,172 +138,226 @@ var EnhanceDns = function (conf) {
...
+                list.push(callback);
+
+                if (list.length > 1) { return; }
+
+                backup_object.resolve6(domain, function (err, addresses) {
+                    if (err) {
+                        list.forEach(function (cb) { cb(err); });
+                        delete callbackList[key];
+                    } else {
+                        cache.set('resolve6_' + domain, addresses, function () {
```
Comment (drewfish) [inline lib/index.js]: Could you use the `key` variable here too? If we change the way the key is constructed it'd be nice to only do it in one place.

File: lib/index.js [hunk 5]
```diff
@@ -126,172 +138,226 @@ var EnhanceDns = function (conf) {
...
+                list.push(callback);
+
+                if (list.length > 1) { return; }
+
+                backup_object.resolveMx(domain, function (err, addresses) {
+                    if (err) {
+                        list.forEach(function (cb) { cb(err); });
+                        delete callbackList[key];
+                    } else {
+                        cache.set('resolveMx_' + domain, addresses, function () {
```
Comment (drewfish) [inline lib/index.js]: Could you use the `key` variable here too? If we change the way the key is constructed it'd be nice to only do it in one place.

File: lib/index.js [hunk 6]
```diff
@@ -126,172 +138,226 @@ var EnhanceDns = function (conf) {
...
+                list.push(callback);
+
+                if (list.length > 1) { return; }
+
+                backup_object.resolveTxt(domain, function (err, addresses) {
+                    if (err) {
+                        list.forEach(function (cb) { cb(err); });
+                        delete callbackList[key];
+                    } else {
+                        cache.set('resolveTxt_' + domain, addresses, function () {
```
Comment (drewfish) [inline lib/index.js]: Could you use the `key` variable here too? If we change the way the key is constructed it'd be nice to only do it in one place.

File: lib/index.js [hunk 7]
```diff
@@ -126,172 +138,226 @@ var EnhanceDns = function (conf) {
...
+                list.push(callback);
+
+                if (list.length > 1) { return; }
+
+                backup_object.resolveSrv(domain, function (err, addresses) {
+                    if (err) {
+                        list.forEach(function (cb) { cb(err); });
+                        delete callbackList[key];
+                    } else {
+                        cache.set('resolveSrv_' + domain, addresses, function () {
```
Comment (drewfish) [inline lib/index.js]: Could you use the `key` variable here too? If we change the way the key is constructed it'd be nice to only do it in one place.

File: lib/index.js [hunk 8]
```diff
@@ -126,172 +138,226 @@ var EnhanceDns = function (conf) {
...
+                list.push(callback);
+
+                if (list.length > 1) { return; }
+
+                backup_object.resolveNs(domain, function (err, addresses) {
+                    if (err) {
+                        list.forEach(function (cb) { cb(err); });
+                        delete callbackList[key];
+                    } else {
+                        cache.set('resolveNs_' + domain, addresses, function () {
```
Comment (drewfish) [inline lib/index.js]: Could you use the `key` variable here too? If we change the way the key is constructed it'd be nice to only do it in one place.

File: lib/index.js [hunk 9]
```diff
@@ -126,172 +138,226 @@ var EnhanceDns = function (conf) {
...
+                list.push(callback);
+
+                if (list.length > 1) { return; }
+
+                backup_object.resolveCname(domain, function (err, addresses) {
+                    if (err) {
+                        list.forEach(function (cb) { cb(err); });
+                        delete callbackList[key];
+                    } else {
+                        cache.set('resolveCname_' + domain, addresses, function () {
```
Comment (drewfish) [inline lib/index.js]: Could you use the `key` variable here too? If we change the way the key is constructed it'd be nice to only do it in one place.

File: lib/index.js [hunk 10]
```diff
@@ -126,172 +138,226 @@ var EnhanceDns = function (conf) {
...
+                list.push(callback);
+
+                if (list.length > 1) { return; }
+
+                backup_object.reverse(ip, function (err, addresses) {
+                    if (err) {
+                        list.forEach(function (cb) { cb(err); });
+                        delete callbackList[key];
+                    } else {
+                        cache.set('reverse_' + ip, addresses, function () {
```
Comment (drewfish) [inline lib/index.js]: Could you use the `key` variable here too? If we change the way the key is constructed it'd be nice to only do it in one place.

**Comment by drewfish** [comment]
Good stuff.

When calling each callback (`list.forEach(function (cb) { cb(...); })`) we might want to call each callback in a separate event loop tick. That way if one of the callbacks throws the other callbacks will still be called. We could call each callback in `nextTick()` or `setImmediate()` or something. (Honestly I don't know too much the difference between them.)

**Comment by jomo** [comment]
You might want to add a warning comment where the `key` variable is assigned / used.

If you change the pattern slightly, e.g to `callbackList[type][domain]` it will fail when the `domain` matches an `Object.prototype` property/method name, such as 'constructor'.

We use the same pattern for concurrent requests in a different project and only noticed when production crashed :cry: https://github.com/crafatar/crafatar/issues/145#issuecomment-143309932

**Comment by drewfish** [comment]
You might also consider using the `inflight` module:
https://www.npmjs.com/package/inflight

**Comment by goloroden** [comment]
Any update on this?

Is there a special reason why this PR was not merged? Is there anything one can do to help?

**Comment by drewfish** [comment]
@goloroden This PR has received a bunch of comments but hasn't been updated, so it appears to be abandoned. (As well there's a merge conflict which needs to be addressed.) If you (or anyone else) wish to submit a different PR which accomplishes the same thing we'd be happy to work to get it merged.

**Comment by goloroden** [comment]
Okay, thanks!

**Comment by goloroden** [comment]
@cloudrac3r Would you mind closing this PR now that #30 has landed?

**Comment by cloudrac3r** [comment]
@goloroden Sorry, what do you want me to do? I am the author of 30, I am not the author of this one.

**Comment by goloroden** [comment]
Sorry, apparently I mixed things up! Didn’t want to bother you 😔

This should probably rather be gone to @drewfish or @esnunes

**Pull request closed without merging**